File: rust/src/graph.rs
extern crate rand;
use std::collections::HashSet;

pub type Vertex = u32;
pub type Adjacencies = Vec<HashSet<Vertex>>;

pub fn new_adjacencies(order: i32) -> Adjacencies {
    std::vec::from_elem(HashSet::new(), order as usize)
}

pub struct UndirectedGraph {
    adjacencies: Adjacencies,
}

impl UndirectedGraph {
    pub fn new(adjacencies: Adjacencies) -> Self {
        for (i, adjacent_to_v) in adjacencies.iter().enumerate() {
            let v = i as Vertex;
            for w in adjacent_to_v {
                assert_ne!(&v, w);
                assert!(
                    adjacencies[*w as usize].contains(&v),
                    format!("{} is adjacent to {} but not vice versa", w, v)
                );
            }
        }
        UndirectedGraph { adjacencies }
    }

    pub fn order(&self) -> i32 {
        self.adjacencies.len() as i32
    }

    pub fn size(&self) -> i32 {
        //total = sum(len(a) for a in self.adjacencies)
        let total: i32 = self.adjacencies.iter().map(|a| a.len() as i32).sum();
        assert!(total % 2 == 0);
        total / 2
    }

    pub fn adjacencies(&self, node: Vertex) -> &HashSet<Vertex> {
        &self.adjacencies[node as usize]
    }
    pub fn degree(&self, node: Vertex) -> i32 {
        self.adjacencies(node).len() as i32
    }

    pub fn connected_nodes(&self) -> HashSet<Vertex> {
        // {node for node in range(self.order) if self.adjacencies[node]}
        (0..self.order() as Vertex)
            .filter(|v| self.degree(*v) > 0)
            .collect()
    }
}

File: python3/graph.py
from typing import List, Set, TypeAlias

Vertex: TypeAlias = int


class UndirectedGraph(object):
    def __init__(self, adjacencies: List[Set[Vertex]]):
        order = len(adjacencies)
        for v, adjacent_to_v in enumerate(adjacencies):
            for w in adjacent_to_v:
                assert 0 <= w < order
                assert v != w, f"{v} is adjacent to itself"
                assert v in adjacencies[w], f"{w} is adjacent to {v} but not vice versa"
        self.adjacencies = adjacencies
        self.max_degree = max(len(a) for a in self.adjacencies) if order else 0
        total_degree = sum(len(a) for a in self.adjacencies)
        assert total_degree % 2 == 0
        self.size = total_degree / 2

    @property
    def order(self) -> int:
        return len(self.adjacencies)

    def degree(self, node: Vertex) -> int:
        return len(self.adjacencies[node])

    def has_degree(self, node: Vertex) -> bool:
        return bool(self.adjacencies[node])

    def connected_vertices(self) -> Set[Vertex]:
        return {node for node in range(self.order) if self.has_degree(node)}
